users-map/build-users-map.sh

```diff
@@ -0,0 +1,109 @@
+  exit 1
+fi
+
+if [ "$#" -lt 2 ]; then
+  error "Usage: $0 /path/to/all-users.csv /path/to/destination.csv"
+  exit 1
+fi
+
+if [ -f "$destination" ]; then
+  error "Destination exists, I won't overwrite it: $destination"
+  exit 1
+fi
+
+search_patterns=$(mktemp --suffix="siretloc.search.txt")
+$CSVTOOL namedcol siret "$all_users_csv" | \
+  # keep only unique values for SIRETs
+  sort -u | \
+  # Discard any unrelevant values (and the header)
+  grep "^[0-9]" | \
+  # Normalize the value, remove the spaces
+  sed "$SED_NORMALIZE_REPLACE" | \
+  # Split the Siret into a tuple of siren and nic, and prefix the pattern with `^` to search only at the beginning of the line
+  sed 's/^\(.\{9\}\)\(.*\)/^\1,\2/g' > "$search_patterns"
+
+stock_etablissement=${STOCK_ETABLISSEMENT:-""}
+if [ "$stock_etablissement" == "" ]; then
+  stock_etablissement_dir=$(mktemp --directory --suffix="stock_etablissement")
+  mkdir -p "$stock_etablissement_dir"
+  stock_etablissement="${stock_etablissement_dir}/StockEtablissementActif_utf8_geo.csv.gz"
+  info "Downloading $stock_etablissement, may take a while"
+  $CURL "https://files.data.gouv.fr/geo-sirene/last/StockEtablissementActif_utf8_geo.csv.gz" > "$stock_etablissement"
+fi
+
+cat="cat"
+if [ "$(file -b --mime-type "$stock_etablissement")" == "application/gzip" ]; then
+  cat="zcat"
+fi
+
+grepped_etablissements_with_loc=$(mktemp --suffix="grepped_etablissements.csv")
+
+cat << EOF | $CSVTOOL namedcol "siret,longitude,latitude" - > "$grepped_etablissements_with_loc"
+$($cat "$stock_etablissement" | head -n 1)
+$($cat "$stock_etablissement" | grep --file="$search_patterns")
+EOF
+
+
+# csvtool join does not seem to work as we would like and seems very complex, let's yolo iterate on the document and grep
+
+siret_col_pos=$(head -n 1 "$all_users_csv" | grep -o "^.*siret" | tr -dc ',' | awk '{ print length + 1; }')
+# Print the header
+echo "$(head -n 1 "$all_users_csv"),longitude,latitude" > "$destination"
+
+while read -r line; do
+  # extract the SIRET from the line, it is at the position $siret_col_pos
+  siret=$(echo "$line" | csvtool col "$siret_col_pos" - | sed "$SED_NORMALIZE_REPLACE")
+  if [ -z "$siret" ]; then
+    echo "$line,," >> "$destination"
+    continue
+  fi
+  # search for the SIRET in the grepped_etablissements_with_loc file and retrieve the location values only
+  loc=$(sed -n "s/^$siret,//p" "$grepped_etablissements_with_loc" || true)
+  if [ -n "$loc" ]; then
+    # if found, append the line to the destination file
+    echo "$line,$loc" >> "$destination"
+  else
+    # if not found, append the line with empty longitude and latitude
+    echo "$line,," >> "$destination"
+  fi
+done < <(tail -n +2 "$all_users_csv") # `tail -n +2` ignores the first line
+
+if [ "${CLEANUP:-0}" == "1" ]; then
+  rm "$search_patterns"
+  rm "$grepped_etablissements_with_loc"
+  if [ -n "${stock_etablissement_dir:-}" ] && [ -d "${stock_etablissement_dir}" ]; then
+    rm -rf "$stock_etablissement_dir"
+  fi
+fi
```

Comment:
One day I found a bash extremist proposing to ban if structure and only use this one

```suggestion
[ "${CLEANUP:-0}" == "1" ] && \
rm "$search_patterns" && \
rm "$grepped_etablissements_with_loc" && \
[ -n "${stock_etablissement_dir:-}" ] && \
[ -d "${stock_etablissement_dir}" ] && \
rm -rf "$stock_etablissement_dir"
```

Not sure it really add more readability, but always a matter of language masterring

Reply:
Hmm, I think it is not very helpful and would make the condition flow less clear.

I prefer the plain `if` :)